nginx/run-benchmark.sh

```diff
@@ -0,0 +1,120 @@
+$(output_options_long)
+EOF
+  exit
+}
+
+parse_params() {
+  TYPE=""
+  THREADS=1
+  DURATION="10s"
+  CONNECTIONS=400
+  NGINX_ITERATIONS=1
+
+  parse_output_params "$@"
+  set -- $UNPARSED
+
+  while [ $# -gt 0 ]; do
+    case "${1-}" in
+    -h | --help) usage ;;
+    -v | --verbose) set -x ;;
+    -t | --threads)
+      THREADS="${2-}"
+      shift
+      ;;
+    -c | --connections)
+      CONNECTIONS="${2-}"
+      shift
+      ;;
+    -d | --duration)
+      DURATION="${2-}"
+      shift
+      ;;
+    -i | --iterations)
+      NGINX_ITERATIONS="${2-}"
+      shift
+      ;;
+    *) ;; # Skip unknown params
+    esac
+    shift
+  done
+}
+
+parse_params "$@"
+
+THREADS=${THREADS}
+CONNECTIONS=${CONNECTIONS}
+DURATION=${DURATION}
+NGINX_ITERATIONS=${NGINX_ITERATIONS}
+PUSHGATEWAY_URL=${PUSHGATEWAY_URL}
+
+SYSTEM=$(/usr/local/bin/cpu.sh system)
+PARAMS="-d $DURATION -c $CONNECTIONS -t $THREADS"
+
+# starts nginx
+echo "Starting nginx"
+nginx
+
+for I in $(seq 1 $NGINX_ITERATIONS); do
+  # Run benchmark and save output to a file
+  wrk ${PARAMS} -s /usr/local/share/wrk/report-json.lua -L http://127.0.0.1:8000/ > /tmp/output-$I.txt
+
+  # Scrape metrics from the file
+  latency_avg=$(cat /tmp/output-$I.txt | grep "Latency Avg" | cut -d":" -f2 | xargs)
+  requests_sec=$(cat /tmp/output-$I.txt | grep "Requests/sec" | cut -d":" -f2 | xargs)
+  transfer_sec=$(cat /tmp/output-$I.txt | grep "Transfer/sec" | cut -d":" -f2 | cut -d"M" -f1 | xargs)
+  total_requests=$(cat /tmp/output-$I.txt | grep "Total Requests" | cut -d":" -f2 | xargs)
+  http_errors=$(cat /tmp/output-$I.txt | grep "HTTP errors" | cut -d":" -f2 | xargs)
+  request_timed_out=$(cat /tmp/output-$I.txt | grep "Requests timed out" | cut -d":" -f2 | xargs)
+  bytes_received=$(cat /tmp/output-$I.txt | grep "Bytes received" | cut -d":" -f2 | xargs)
+  socket_connect_errors=$(cat /tmp/output-$I.txt | grep "Socket connect errors" | cut -d":" -f2 | xargs)
+  socket_read_errors=$(cat /tmp/output-$I.txt | grep "Socket read errors" | cut -d":" -f2 | xargs)
+  socket_write_errors=$(cat /tmp/output-$I.txt | grep "Socket write errors" | cut -d":" -f2 | xargs)
+
+  cat <<EOF | curl --data-binary @- $PUSHGATEWAY_URL/nginx-$I
+
+    # TYPE latency_avg gauge
+    $(echo "latency_avg ${latency_avg}")
+
+    # TYPE requests_sec gauge
+    $(echo "requests_sec ${requests_sec}")
+
+    # TYPE transfer_sec gauge
+    $(echo "transfer_sec ${transfer_sec}")
+
+    # TYPE total_requests gauge
+    $(echo "total_requests ${total_requests}")
+
+    # TYPE http_errors gauge
+    $(echo "http_errors ${http_errors}")
+
+    # TYPE request_timed_out gauge
+    $(echo "request_timed_out ${request_timed_out}")
+
+    # TYPE bytes_received gauge
+    $(echo "bytes_received ${bytes_received}")
+
+    # TYPE socket_ gauge
+    $(echo "socket_connect_errors ${socket_connect_errors}")
+    $(echo "socket_read_errors ${socket_read_errors}")
+    $(echo "socket_write_errors ${socket_write_errors}")
```

Comment:
Also I don't see all these metrics, I see only following two:

![Screenshot from 2021-04-30 18-40-18](https://user-images.githubusercontent.com/5815795/116699734-8d67ea00-a9e3-11eb-93e8-c1d72421862c.png)

Reply:
Also from the logs I got this error:

```console
$ kl run-nginx-benchmark-v95sz
Starting nginx
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
pushed metrics are invalid or inconsistent with existing metrics: collected metric "latency_avg" { label:<name:"instance" value:"nginx-1" > label:<name:"job" value:"nginx-benchmark-job" > gauge:<value:2 > } is not a UNTYPED
100   722  100   224  100   498  44800  99600 --:--:-- --:--:-- --:--:--  141k

```

Reply:
Fixed